x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java

```diff
@@ -392,41 +392,31 @@ public static Query doTranslate(In in, TranslatorHandler handler) {
         }
 
         private static Query translate(In in, TranslatorHandler handler) {
-            Query q;
-            if (in.value() instanceof FieldAttribute) {
-                // equality should always be against an exact match (which is important for strings)
-                FieldAttribute fa = (FieldAttribute) in.value();
-                DataType dt = fa.dataType();
-
-                List<Expression> list = in.list();
-                Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(list.size()));
-                list.forEach(e -> {
-                    // TODO: this needs to be handled inside the optimizer
-                    if (DataTypes.isNull(e.dataType()) == false) {
-                        set.add(handler.convert(valueOf(e), dt));
-                    }
-                });
-
-                if (DataTypes.isDateTime(dt)) {
-                    DateFormatter formatter = DateFormatter.forPattern(DATE_FORMAT);
-
-                    q = null;
-                    for (Object o : set) {
-                        assert o instanceof ZonedDateTime : "expected a ZonedDateTime, but got: " + o.getClass().getName();
-                        // see comment in Ranges#doTranslate() as to why formatting as String is required
-                        String zdt = formatter.format((ZonedDateTime) o);
-                        RangeQuery right = new RangeQuery(
-                            in.source(), fa.exactAttribute().name(),
-                            zdt, true, zdt, true, formatter.pattern(), in.zoneId());
-                        q = q == null ? right : new BoolQuery(in.source(), false, q, right);
+            assert in.value() instanceof FieldAttribute;
+
+            Set<Object> terms = new LinkedHashSet<>();
+            List<Query> queries = new ArrayList<>();
+
+            for (Expression rhs : in.list()) {
+                if (DataTypes.isNull(rhs.dataType()) == false) {
+                    // delegates to BinaryComparisons translator to ensure consistent handling of date and time values
+                    Query query = BinaryComparisons.translate(new Equals(in.source(), in.value(), rhs, in.zoneId()), handler);
+
+                    if (query instanceof TermQuery) {
+                        terms.add(((TermQuery) query).value());
+                    } else {
+                        queries.add(query);
                     }
-                } else {
-                    q = new TermsQuery(in.source(), fa.exactAttribute().name(), set);
                 }
-            } else {
-                q = new ScriptQuery(in.source(), in.asScript());
```

Comment:
Is my understanding correct that the `ScriptQuery` is now created in the `BinaryComparison.translate`?

Reply:
This branch was never reached because `in.value()` is always a `FieldAttribute`. The script query is always created by `handler.wrapFunctionQuery` in `doTranslate`.